reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,382 @@
+            this.callMode = callMode;
+            this.ordering = ordering;
+            this.acyclicCost = acyclicCost;
+            this.cyclicConcludableModes = cyclicConcludableModes;
+            this.scalingFactors = scalingFactors;
+        }
+    }
+
+    static class PartialOrderReductionSearch {
+        private final Set<Resolvable<?>> resolvables;
+        private final Set<Variable> mode;
+        private final Map<Resolvable<?>, Set<Variable>> dependencies;
+        private final boolean enableConnectednessRestriction;
+
+        // TODO: Combine these when we fix ReasonerPlanner.dependencies
+        PartialOrderReductionSearch(Set<Resolvable<?>> resolvables, Set<Variable> mode) {
+            this(resolvables, mode, ReasonerPlanner.dependencies(resolvables), true);
+        }
+
+        PartialOrderReductionSearch(Set<Resolvable<?>> resolvables, Set<Variable> mode, Map<Resolvable<?>, Set<Variable>> dependencies, boolean enableConnectednessRestriction) {
+            this.resolvables = resolvables;
+            this.mode = mode;
+            this.dependencies = dependencies;
+            this.enableConnectednessRestriction = enableConnectednessRestriction;
+        }
+
+        List<List<Resolvable<?>>> allOrderings() {
+            Set<Resolvable<?>> remaining = new HashSet<>(resolvables);
+            Set<Variable> bounds = new HashSet<>(mode);
+            iterate(resolvables).filter(resolvable -> !resolvable.isNegated()).flatMap(r -> iterate(estimateableVariables(r)))
+                    .filter(v -> iterate(v.constraints()).anyMatch(constraint ->
+                            constraint.isThing() && constraint.asThing().isValue() && constraint.asThing().asValue().isValueIdentity()))
+                    .forEachRemaining(bounds::add);
+
+            List<List<Resolvable<?>>> orderings = new ArrayList<>();
+            porDFS(new Stack<>(), bounds, remaining, new HashSet<>(), orderings);
+            return orderings; // TODO: If this causes a memory blow-up, POR can "generate" one at a time.
+        }
+
+        private void porDFS(Stack<Resolvable<?>> currentPath, Set<Variable> currentBounds,
+                            Set<Resolvable<?>> remaining, Set<Resolvable<?>> sleepSet, List<List<Resolvable<?>>> orderings) {
+            if (remaining.isEmpty()) {
+                orderings.add(new ArrayList<>(currentPath));
+                return;
+            }
+
+            List<Resolvable<?>> enabled = iterate(remaining)
+                    .filter(r -> ReasonerPlanner.dependenciesSatisfied(r, currentBounds, dependencies) && !sleepSet.contains(r))
+                    .toList();
+
+            if (enabled.isEmpty()) {
+                enabled = iterate(remaining).filter(r -> !r.isNegated() && !sleepSet.contains(r)).toList();
+            }
+
+            if (enabled.isEmpty()) {
+                return; // All enabled are sleeping
+            }
+
+            if (enableConnectednessRestriction) {
```

Comment:
this is interesting! What's this boolean and why do we have it?

Reply:
Reduces the search-space by only scheduling resolvables which are connected to some bound.
We can get rid of it and make it the default behaviour, because I think the alternative will just be too slow.

Reply:
okay, let's do that